trame_server/utils/state_proxy.py

```diff
@@ -0,0 +1,420 @@
+    """
+    Converts the input state proxy instance to dataclass.
+    """
+    dataclass_type = _get_proxy_dataclass_type_or_raise(instance)
+
+    kwargs = {}
+    for f in fields(dataclass_type):
+        attr = getattr(instance, f.name)
+        if _is_proxy_class(attr):
+            kwargs[f.name] = as_dataclass(attr)
+        else:
+            kwargs[f.name] = attr
+
+    return dataclass_type(**kwargs)
+
+
+def from_dataclass(instance: T, dataclass_obj: T) -> None:
+    """
+    Populate the state proxy instance from the values of the given dataclass object.
+    """
+    dataclass_type = _get_proxy_dataclass_type_or_raise(instance)
+
+    if not isinstance(dataclass_obj, dataclass_type):
+        _error_msg = f"Expected instance of {dataclass_type.__name__}, got {type(dataclass_obj).__name__}"
+        raise TypeError(_error_msg)
+
+    for f in fields(dataclass_type):
+        attr = getattr(instance, f.name)
+        value = getattr(dataclass_obj, f.name)
+
+        if _is_proxy_class(attr):
+            from_dataclass(attr, value)
+        else:
+            setattr(instance, f.name, value)
+
+
+def get_field_proxy_dict(instance: T) -> dict[str, ProxyField]:
+    """
+    :returns: State ID to ProxyField as saved in the input instance.
+    """
+    _get_proxy_dataclass_type_or_raise(instance)
+    return getattr(instance, _STATE_PROXY_FIELD_DICT, {})
+
+
+def bind_typed_state_change(
+    keys: Any | tuple[Any], callback: Callable, state: State, data: T
+) -> None:
+    """
+    Bind state id changes to callbacks.
+    Callbacks are called with converted state values found with input keys.
+    """
+    state_id_to_field_dict = get_field_proxy_dict(data)
+
+    if isinstance(keys, list):
+        keys = tuple(keys)
+
+    if not isinstance(keys, tuple):
+        keys = (keys,)
+
+    @state.change(*keys)
+    def _on_state_change(**_):
+        values = [state_id_to_field_dict[k].get_value() for k in keys]
+        callback(*values)
+
+
+class TypedState(Generic[T]):
+    """
+    Helper to have access to, mutate, and be notified of state changes using a strongly typed dataclass interface.
+    """
+
+    def __init__(
+        self,
+        state: State,
+        dataclass_type: Type[T],
+        *,
+        state_namespace="",
+        state_encoder: IStateEncoderDecoder | None = None,
+    ):
+        self._state = state
+
+        self.data, self.name = state_proxies(
+            dataclass_type,
+            state,
+            state_namespace=state_namespace,
+            state_encoder=state_encoder,
+        )
+
+    @overload
+    def bind_change(self, key: Any, callback: Callable) -> None: ...
+
+    @overload
+    def bind_change(self, key: tuple[Any], callback: Callable) -> None: ...
+
+    @overload
+    def bind_change(self, change_dict: dict[Any | list[Any], Callable]) -> None: ...
+
+    def bind_change(self, *args) -> None:
+        if len(args) == 1 and isinstance(args[0], dict):
```

Comment:
you may want to add some docstring to explain usage especially since you have a couple of way to call it.

Reply:
Yes, sorry about that, I pushed a first API example to keep discussing implementation and API here but its missing docstrings.

Let me know how / where I should document the proxy mechanism as well.
I will add the necessary documentation once we agree on how it should work.